File: Walmart.Sgp.WebApp/Scripts/app/services/notaFiscal.service.js
(function () {
    'use strict';

    angular
        .module('SGP')
        .service('NotaFiscalService', NotaFiscalService);

    NotaFiscalService.$inject = ['$q', '$http', 'ApiEndpoints'];

    function NotaFiscalService($q, $http, ApiEndpoints) {

        function carregaNotaFiscal(response) {
            return response.data;
        }

        function carregaNotaFiscalItemCustos(response) {
            return response.data;
        }

        function carregaUltimasEntradasNotaFiscal(response) {
            return response.data;
        }

        function existeNotasPendentes(response) {
            return response.data;
        }

        var self = this;

        self.obterEstruturado = function (id) {

            return $http
                .get(ApiEndpoints.sgp.notaFiscal + ApiEndpoints.encode(id) + '/Estruturado')
                .then(carregaNotaFiscal);
        };

        self.obterItensDaNotaFiscal = function (id, paging) {
            var params = ApiEndpoints.createParams(paging, {});

            return $http
                .get(ApiEndpoints.sgp.notaFiscal + ApiEndpoints.encode(id) + '/Itens', { params: params })
                .then(carregaNotaFiscal);
        };

        self.obterCustosPorItem = function (cdLoja, cdItem, dtSolicitacao) {

            return $http
                .get(ApiEndpoints.sgp.notaFiscal + 'ObterCustosPorItem', { params: { cdLoja: cdLoja, cdItem: cdItem, dtSolicitacao: dtSolicitacao } })
                .then(carregaNotaFiscalItemCustos);
        };

        self.existeNotasPendentesPorItem = function (cdLoja, cdItem, dtSolicitacao) {

            return $http
                .get(ApiEndpoints.sgp.notaFiscal + 'NotasPendentes', { params: { cdLoja: cdLoja, cdItem: cdItem, dtSolicitacao: dtSolicitacao } })
                .then(existeNotasPendentes);
        };

        self.pesquisarPorFiltros = function (filtro, paging) {
            var params = ApiEndpoints.createParams(paging, {
                cdSistema: filtro.cdSistema,
                idBandeira: filtro.idBandeira,
                cdLoja: filtro.cdLoja,
                cdFornecedor: filtro.cdFornecedor,
                nrNotaFiscal: filtro.nrNotaFiscal,
                cdItem: filtro.cdItem,
                dtRecebimento: filtro.dtRecebimento,
                dtCadastroConcentrador: filtro.dtCadastroConcentrador,
                dtAtualizacaoConcentrador: filtro.dtAtualizacaoConcentrador
            });
                        
            return $http
                .get(ApiEndpoints.sgp.notaFiscal + 'PesquisarPorFiltros', { params: JSON.flatten(params) })
                .then(carregaNotaFiscal);
        };

        self.pesquisarUltimasEntradasPorFiltro = function (idItemDetalhe, idLoja, dtSolicitacao, paging) {

            var params = ApiEndpoints.createParams(paging, {
                idItemDetalhe: idItemDetalhe,
                idLoja: idLoja,
                dtSolicitacao: dtSolicitacao
            });

            return $http
                .get(ApiEndpoints.sgp.notaFiscal + 'PesquisarUltimasEntradas', { params: params })
                .then(carregaUltimasEntradasNotaFiscal);
        };

        self.pesquisarCustosPorFiltros = function (filtro, paging) {

            var params = ApiEndpoints.createParams(paging, {
                idBandeira: filtro.idBandeira, 
                idLoja: !filtro.loja ? null : filtro.loja.idLoja,
                idFornecedor: !filtro.fornecedor ? null : filtro.fornecedor.idFornecedor,
                nrNotaFiscal: filtro.nrNotaFiscal,
                idItemDetalhe: !filtro.item ? null : filtro.item.idItemDetalhe,
                idDepartamento: !filtro.departamento ? null : filtro.departamento.idDepartamento,
                idNotaFiscalItemStatus: filtro.idNotaFiscalItemStatus,          
                dtRecebimento: filtro.dtRecebimento,
                dtCadastroConcentrador: filtro.dtCadastroConcentrador,
                dtAtualizacaoConcentrador: filtro.dtAtualizacaoConcentrador
            });

            return $http
                .get(ApiEndpoints.sgp.notaFiscal + 'PesquisarCustosPorFiltros', { params: JSON.flatten(params) })
                .then(carregaNotaFiscal);
        };

        self.corrigirCustos = function (custos) {

            var itens = [];

            angular.forEach(custos, function (value) {
                // TODO: talvez seja possível remover o idBandeira
                itens.push({
                    idNotaFiscalItem: value.idNotaFiscalItem,
                    idBandeira: value.idBandeira,
                    blLiberar: value.blLiberar,
                    qtAjustada: value.qtAjustada
                });
            });

            return $http
                .post(ApiEndpoints.sgp.notaFiscal + 'CorrigirCustos', itens);
        };
    }
})();

File: Walmart.Sgp.Domain/Movimentacao/NotaFiscalService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Walmart.Sgp.Domain.Item;
using Walmart.Sgp.Infrastructure.Framework.Domain;
using Walmart.Sgp.Infrastructure.Framework.Specs;

namespace Walmart.Sgp.Domain.Movimentacao
{
    /// <summary>
    /// Serviço de domínio relacionado a nota fiscal.
    /// </summary>
    public class NotaFiscalService : DomainServiceBase<INotaFiscalGateway>, INotaFiscalService
    {
        #region Fields
        private readonly INotaFiscalItemGateway m_notaFiscalItemGateway;
        #endregion

        #region Constructors
        /// <summary>
        /// Inicia uma nova instância da classe <see cref="NotaFiscalService"/>.
        /// </summary>
        /// <param name="notaFiscalGateway">O table data gateway para nota fiscal.</param>
        /// <param name="notaFiscalItemGateway">O table data gateway para item de nota fiscal.</param>
        public NotaFiscalService(INotaFiscalGateway notaFiscalGateway, INotaFiscalItemGateway notaFiscalItemGateway)
            : base(notaFiscalGateway)
        {
            m_notaFiscalItemGateway = notaFiscalItemGateway;
        }
        #endregion

        #region Methods
        /// <summary>
        /// Obtém uma nota fiscal pelo seu id e retorna a nota fiscal com informações das entidades associadas.
        /// </summary>
        /// <param name="idNotaFiscal">O id de nota fiscal.</param>
        /// <returns>A NotaFiscal com informações de Loja, Fornecedor e Bandeira.</returns>
        public NotaFiscal ObterEstruturadoPorId(int idNotaFiscal)
        {
            return this.MainGateway.ObterEstruturadoPorId(idNotaFiscal);
        }

        /// <summary>
        /// Pesquisa detalhe de notas fiscais pelos filtros informados.
        /// </summary>
        /// <param name="filtro">O filtro</param>
        /// <param name="paging">A paginação que será utilizada para retornar o resultado.</param>
        /// <returns>Os itens.</returns>
        public IEnumerable<NotaFiscal> PesquisarPorFiltros(NotaFiscalFiltro filtro, Paging paging)
        {
            Assert(new { dtRecebimento = filtro.DtRecebimento }, new AllMustBeInformedSpec());

            return this.MainGateway.PesquisarPorFiltros(filtro, paging);
        }

        /// <summary>
        /// Pesquisa de custos de notas fiscais pelos filtros informados.
        /// </summary>
        /// <param name="filtro">O filtro.</param>
        /// <param name="paging">A paginação que será utilizada para retornar o resultado.</param>
        /// <returns>Os itens</returns>
        public IEnumerable<CustoNotaFiscal> PesquisarCustosPorFiltros(NotaFiscalFiltro filtro, Paging paging)
        {
            Assert(new { dtRecebimento = filtro.DtRecebimento }, new AllMustBeInformedSpec());

            return this.MainGateway.PesquisarCustosPorFiltros(filtro, paging);
        }

        /// <summary>
        /// Obtém o último item de nota fiscal recebido na loja para o item detalhe informado.
        /// </summary>
        /// <param name="idItemDetalhe">O id do item detalhe.</param>
        /// <param name="idLoja">O id da loja.</param>
        /// <returns>O item na última nota fiscal da loja.</returns>
        public NotaFiscalItem ObterItemNaUltimaNotaRecebidaDaLoja(int idItemDetalhe, int idLoja)
        {
            return MainGateway.ObterItemNaUltimaNotaRecebidaDaLoja(idItemDetalhe, idLoja);
        }

        /// <summary>
        /// Obtém os itens da nota fiscal com o id informado.
        /// </summary>
        /// <param name="idNotaFiscal">O id da nota fiscal.</param>
        /// <param name="paging">A paginação que será utilizada para retornar o resultado.</param>
        /// <returns>Os itens da nota fiscal.</returns>        
        public IEnumerable<NotaFiscalItem> ObterItensDaNotaFiscal(int idNotaFiscal, Paging paging)
        {
            return MainGateway.ObterItensDaNotaFiscal(idNotaFiscal, paging);
        }

        /// <summary>
        /// Pesquisa ultimas entradas de notas fiscais pelos filtros informados.
        /// </summary>
        /// <param name="idItemDetalhe">O ID do item</param>
        /// <param name="idLoja">O ID da loja</param>        
        /// <param name="dtSolicitacao">A data da solicitacao</param>
        /// <param name="paging">A paginacao</param>
        /// <returns>As entradas</returns>
        public IEnumerable<NotaFiscalConsolidado> PesquisarUltimasEntradasPorFiltros(long idItemDetalhe, int idLoja, DateTime dtSolicitacao, Paging paging)
        {
            Assert(new { dtSolicitacao = dtSolicitacao }, new AllMustBeInformedSpec());

            return this.MainGateway.PesquisarUltimasEntradasPorFiltro(idItemDetalhe, idLoja, dtSolicitacao, paging);
        }

        /// <summary>
        /// Pesquisa os custos do item.
        /// </summary>
        /// <param name="cdLoja">O codigo da loja</param>
        /// <param name="cdItem">O codigo do item</param>
        /// <param name="dtSolicitacao">A data da solicitacao</param>
        /// <returns>Os custos do item.</returns>
        public NotaFiscalItemCustosConsolidado ObterCustosPorItem(int cdLoja, long cdItem, DateTime dtSolicitacao)
        {
            return this.MainGateway.ObterCustosPorItem(cdLoja, cdItem, dtSolicitacao);
        }

        /// <summary>
        /// Verifica se existe notas pendentes para o item informado
        /// </summary>
        /// <param name="cdLoja">O codigo da loja</param>
        /// <param name="cdItem">O codigo do item</param>
        /// <param name="dtSolicitacao">A data da solicitacao</param>
        /// <returns>Verdadeiro ou falso</returns>
        public bool ExisteNotasPendentesPorItem(int cdLoja, long cdItem, DateTime dtSolicitacao)
        {
            return this.MainGateway.ExisteNotasPendentesPorItem(cdLoja, cdItem, dtSolicitacao);
        }

        /// <summary>
        /// Corrige os custos de notas fiscais.
        /// </summary>
        /// <param name="custos">Os custos.</param>
        public void CorrigirCustos(IEnumerable<CustoNotaFiscal> custos)
        {
            var custosLiberados = custos.Where(o => o.blLiberar).ToList();

            foreach (var custo in custosLiberados)
            {
                var notaFiscalItem = m_notaFiscalItemGateway.FindById(custo.IDNotaFiscalItem);

                AtualizarCorrecaoCustoNotaFiscalItem(custo, notaFiscalItem);

                m_notaFiscalItemGateway.Update(notaFiscalItem);
                m_notaFiscalItemGateway.SalvarLogCorrecaoCusto(notaFiscalItem);
                m_notaFiscalItemGateway.LiberarItemNotaFiscalDivergente(notaFiscalItem, custo.IDBandeira);
            }
        }

        private static void AtualizarCorrecaoCustoNotaFiscalItem(CustoNotaFiscal custo, NotaFiscalItem notaFiscalItem)
        {
            notaFiscalItem.dtLiberacao = DateTime.UtcNow;
            notaFiscalItem.qtItemAnterior = notaFiscalItem.qtItem;
            notaFiscalItem.blDivergente = false;

            AlterarStatusNotaFiscalItem(custo, notaFiscalItem);

            notaFiscalItem.qtItem = custo.qtAjustada;

            if (custo.qtAjustada != 0)
            {
                notaFiscalItem.vlCusto = notaFiscalItem.vlMercadoria / custo.qtAjustada;
            }
        }

        private static void AlterarStatusNotaFiscalItem(CustoNotaFiscal custo, NotaFiscalItem notaFiscalItem)
        {
            if (notaFiscalItem.IdNotaFiscalItemStatus != NotaFiscalItemStatus.IdConforme && custo.qtAjustada != notaFiscalItem.qtItem)
            {
                notaFiscalItem.IdNotaFiscalItemStatus = NotaFiscalItemStatus.IdAlterado;
            }
            else if (notaFiscalItem.IdNotaFiscalItemStatus != NotaFiscalItemStatus.IdConforme && custo.qtAjustada == notaFiscalItem.qtItem)
            {
                notaFiscalItem.IdNotaFiscalItemStatus = NotaFiscalItemStatus.IdRevisado;
            }
        }
        #endregion
    }
}
